src/entry.c

```
@@ -0,0 +1,356 @@
+             *     | TTL  | field        | value         |
+             *     |      | hdr "foo" \0 | hdr8 "bar" \0 |
+             *     +------+--------------+---------------+
+             */
+            embed_value = true;
+            alloc_size += embedded_value_size;
+        } else {
+            /* Embed field, but not value. Field must be >= SDS_TYPE_8 to encode to
+             * indicate this type of entry.
+             *
+             *     +------+-------+---------------+
+             *     | TTL  | value | field         |
+             *     |      | ptr   | hdr8 "foo" \0 |
+             *     +------+-------+---------------+
+             */
+            embed_value = false;
+            alloc_size += sizeof(sds);
+            if (field_sds_type == SDS_TYPE_5) {
+                field_sds_type = SDS_TYPE_8;
+                alloc_size -= field_size;
+                field_size = sdsReqSize(field_len, field_sds_type);
+                alloc_size += field_size;
+            }
+        }
+    }
+    /* allocate the buffer */
+    size_t buf_size;
+    char *buf = zmalloc_usable(alloc_size, &buf_size);
+
+    /* Set The expiry if exists */
+    if (expiry_size) {
+        memcpy(buf, &expiry, expiry_size);
+        buf += expiry_size;
+        buf_size -= expiry_size;
+    }
+    if (value) {
+        if (!embed_value) {
+            *(sds *)buf = value;
+            buf += sizeof(sds);
+            buf_size -= sizeof(sds);
+        } else {
+            sdswrite(buf + field_size, buf_size - field_size, SDS_TYPE_8, value, value_len);
+            sdsfree(value);
+            buf_size -= embedded_value_size;
+        }
+    }
+    /* Set the field data */
+    embedded_field_sds = sdswrite(buf, field_size, field_sds_type, field, field_len);
+
+    /* Field sds aux bits are zero, which we use for this entry encoding. */
+    sdsSetAuxBit(embedded_field_sds, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR, embed_value ? 0 : 1);
```

Is there a reason we don't assert when sdsSetAuxBit is not called on SDS_TYPE_5? It feels like that is definitely a coding bug, and instead we just silently return?

---

Actually field sds of type5 is definitely possible when you get here. so you suggest something like?:
```
sdsSetAuxBit(embedded_field_sds, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR, embed_value ? 0 : 1);
sdsSetAuxBit(embedded_field_sds, FIELD_SDS_AUX_BIT_ENTRY_HAS_EXPIRY, expiry_size > 0 ? 1 : 0);

// New asserts
serverAssert(sdsGetAuxBin(embedded_field_sds, FIELD_SDS_AUX_BIT_ENTRY_HAS_VALUE_PTR) == embed_value ? 0 : 1);
serverAssert(sdsGetAuxBin(embedded_field_sds, FIELD_SDS_AUX_BIT_ENTRY_HAS_EXPIRY) == expiry_size > 0 ? 1 : 0);
``` 

If so, I had these asserts for a long time during development when I changed the entry code a bit, but now it felt like very straight forward. I can present them back if you feel it could do good (I prefer in debugAssert though)

---

> Is there a reason we don't assert when sdsSetAuxBit is not called on SDS_TYPE_5? 

+1. 

@ranshid I think Madelyn's comment is on `sdsSetAuxBit`. overall I feel that this function need a lot more debug or server asserts. `sdsSetAuxBit` not having any boundary checks today can definitely hide bugs.

---

I can place these asserts back. BTW we do have unittests for the entry code and most (if not all) these cases are being tested there